src/cloudformation_cli_python_lib/resource.py

```diff
@@ -196,22 +192,47 @@ def _parse_request(
                 aws_secret_access_key=platform_creds.secretAccessKey,
                 aws_session_token=platform_creds.sessionToken,
             )
+            provider_sess = (
+                boto3.Session(
+                    aws_access_key_id=provider_creds.accessKeyId,
+                    aws_secret_access_key=provider_creds.secretAccessKey,
+                    aws_session_token=provider_creds.sessionToken,
+                )
+                if provider_creds
+                else None
+            )
             action = Action[event.action]
             callback_context = event.requestContext.get("callbackContext", {})
-            Credentials(**event_data["requestData"]["platformCredentials"])
         except Exception as e:  # pylint: disable=broad-except
             LOG.exception("Invalid request")
             raise InvalidRequest(f"{e} ({type(e).__name__})") from e
-        return caller_sess, platform_sess, request, action, callback_context, event
+        return (
+            (caller_sess, provider_sess, platform_sess),
+            request,
+            action,
+            callback_context,
+            event,
+        )
 
-    @_ensure_serialize
-    def __call__(
+    # TODO: refactor to reduce branching and locals
+    @_ensure_serialize  # noqa: C901
+    def __call__(  # pylint: disable=too-many-locals
         self, event_data: MutableMapping[str, Any], context: LambdaContext
     ) -> MutableMapping[str, Any]:
         try:
             ProviderLogHandler.setup(event_data)
-            parsed = self._parse_request(event_data)
-            caller_sess, platform_sess, request, action, callback, event = parsed
+            sessions, request, action, callback, event = self._parse_request(event_data)
+            caller_sess, provider_sess, platform_sess = sessions
+            metrics = MetricsPublisherProxy()
+            metrics.add_metrics_publisher(
+                MetricPublisher(event.awsAccountId, event.resourceType, platform_sess)
+            )
+            if provider_sess:
+                metrics.add_metrics_publisher(
+                    MetricPublisher(
+                        event.awsAccountId, event.resourceType, provider_sess
+                    )
+                )
```

Comment:
this looks like a case to me where breaking from the Java interface could be beneficial? we could backport changes or simply say it's different; the metrics stuff is old and the log delivery was patched in hastily from what i can see. for example, seems like the only difference between the publishers is the session. that means the proxy could take the `awsAccountId` and `resourceType`, and maybe `add_metrics_publisher` would just be called with credentials to add another one?

it's up to you. either one of us could do this in a separate PR, too. more of a brainstorm, since it's perfectly functional as is.

Reply:
let's get this merged to get the functionality available and can improve in future pr's